EIP-2980: Swiss Compliant Asset Token (#2980)
ethereum/EIPs: EIPS/eip-erc20-swiss-compl.md

---
eip: <to be assigned>
title: ERC-20-compatible Swiss Compliant Asset Token
author: Gianluca Perletti (@Perlets9), Alan Scarpellini (@alanscarpellini), Roberto Gorini (@robertogorini), Manuel Olivi (@manvel79)
discussions-to: Pull Request
status: Draft
type: Standards Track
category: ERC
created: 2020-09-08
requires: 20
---

## Simple Summary

> An interface for asset tokens (or security tokens), compliant with Swiss Law and compatible with ERC-20.

Reviewer: ```suggestion
An interface for asset tokens (or security tokens), compliant with Swiss Law and compatible with [ERC-20](./eip-20.md)
```

## Abstract

> This new standard is an ERC-20 compatible token with restrictions that comply with one or more one the following Swiss laws: Stock Exchange Act, the Banking Act, the Financial Market Infrastructure Act, the Act on Collective Investment Schemes and the Anti-Money Laundering Act. The Financial Services Act and the Financial Institutions Act must also be considered. The solution achieved meet also the European jurisdiction.

Reviewer: ```suggestion
This new standard is an [ERC-20](./eip-20.md) compatible token with restrictions that comply with one or more one the following Swiss laws: Stock Exchange Act, the Banking Act, the Financial Market Infrastructure Act, the Act on Collective Investment Schemes and the Anti-Money Laundering Act. The Financial Services Act and the Financial Institutions Act must also be considered. The solution achieved meet also the European jurisdiction.
```

This new standard meets the new era of asset tokens (or security tokens). These new methods manage securities ownership during issuance and trading. The issuer is the only role that can manage a white-listing and the only one that is allowed to execute “freeze” or “revoke” functions.

## Motivation

In its ICO guidance dated February 16, 2018, FINMA (Swiss Financial Market Supervisory Authority) defines asset tokens as tokens representing assets and/or relative rights. It explicitly mentions that asset tokens are analogous to and can economically represent shares, bonds, or derivatives. The long list of relevant financial market laws mentioned above reveal that we need more methods than with Payment and Utility Token.

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.

The words "asset tokens" and "security tokens" can be considered synonymous.

Every ERC-toBeAssigned compliant contract must implement the ERCtoBeAssigned interface.

### ERCtoBeAssigned (Token Contract)

``` solidity

> interface ERCtoBeAssigned {

Reviewer: ```suggestion
interface ERC2980 extends ERC20 {
```

  
  /// @dev This emits when funds are reassigned
  event FundsReassigned(address from, address to, uint256 amount);

  /// @dev This emits when funds are revoked
  event FundsRevoked(address from, uint256 amount);

  /// @dev This emits when an address is frozen
  event FundsFrozen(address target);

>   /**
>   * @dev Transfer tokens from a specified address to another one
>   * this operation can be performmed only by an Issuer.
>   * @param _from The address from which the tokens are withdrawn
>   * @param _to The address that receives the tokens
>   * @return true if the tokens are transferred
>   */
>   function reassign(address _from, address _to) external;
> 

Reviewer: ```suggestion
```

Any functions that are issuer only shouldn't be part of the specification since there isn't a many-to-many relationship between tokens and users.  Essentially, this is an implementation detail and not part of the standardized interface.

In this case, for any given token there will be one issuer and that issuer can simply use tools that work with that specific token.  For example, if one token names this function `reassign` and another token names this function `steal`, both tokens can work and general purpose tools can still work with them.  Each issuer would simply use the tooling for that specific token, or design the token around a particular tool that they like.

>   /**
>   * @dev Transfer tokens from a specified address to the Issuer who invokes the method
>   * this operation can be performmed only by an Issuer.
>   * @param _from The address from which the tokens are withdrawn
>   * @return true if the tokens are transferred
>   */
>   function revoke(address _from) external;
> 

Reviewer: ```suggestion
```
Same comment as above, issuer only functions shouldn't be part of the standard.

  /**
  * @dev getter to determine if address is in frozenlist
  */
  function frozenlist(address _operator) external view returns (bool);

>   /**
>   * @dev add an address to the frozenlist
>   * this operation can be performmed only by an Issuer.
>   * @param _operator address
>   * @return true if the address was added to the frozenlist, false if the address was already in the frozenlist
>   */
>   function addAddressToFrozenlist(address _operator) external;
> 

Reviewer: ```suggestion
```
Issuer only, remove.

>   /**
>   * @dev remove an address from the frozenlist
>   * this operation can be performmed only by an Issuer.
>   * @param _operator address
>   * @return true if the address was removed from the frozenlist,
>   * false if the address wasn't in the frozenlist in the first place
>   */
>   function removeAddressFromFrozenlist(address _operator) external;
> 

Reviewer: ```suggestion
```
Issuer only, remove.

  /**
  * @dev getter to determine if address is in whitelist
  */
  function whitelist(address _operator) external view returns (bool);

>   /**
>   * @dev add an address to the whitelist
>   * this operation can be performmed only by an Issuer.
>   * @param _operator address
>   * @return true if the address was added to the whitelist, false if the address was already in the whitelist
>   */
>   function addAddressToWhitelist(address _operator) external;
> 

Reviewer: ```suggestion
```
Issuer only, remove.

>   /**
>   * @dev remove an address from the whitelist
>   * this operation can be performmed only by an Issuer.
>   * @param _operator address
>   * @return true if the address was removed from the whitelist,
>   * false if the address wasn't in the whitelist in the first place
>   */
>   function removeAddressFromWhitelist(address _operator) external;
> 

Reviewer: ```suggestion
```
Issuer only, remove.

>   /**
>   * @dev add a new issuer address
>   * this operation can be performmed only by the contract Owner.
>   * @param _operator address
>   * @return true if the address was not an issuer, false if the address was already an issuer
>   */
>   function addIssuer(address _operator) external;
> 

Reviewer: ```suggestion
```
Owner only, remove.

  /**
  * @dev remove an address from issuers
  * this operation can be performmed only by the contract Owner.
  * @param _operator address
  * @return true if the address has been removed from issuers,
  * false if the address wasn't in the issuer list in the first place
  */
  function removeIssuer(address _operator) external;

  /**
  * @dev Allows the current issuer to transfer his role to a newIssuer.
  * this operation can be performmed only by an Issuer.
  * @param _newIssuer The address to transfer the issuer role to.
  */
  function transferIssuer(address _newIssuer) external;

}
```

> The ERCtoBeAssigned extends ERC-20. Due to the indivisible nature of asset tokens, the decimals number MUST be zero.

Reviewer: ```suggestion
The ERC-2980 extends [ERC-20](./eip-20.md). Due to the indivisible nature of asset tokens, the decimals number MUST be zero.
```

### Revoke and Reassign

Revoke and Reassign methods allow Issuers to move tokens from addresses, even if they are in the Frozenlist. The Revoke method transfers the entire balance of the target address to the Issuer who invoked the method. The Reassign method transfers the entire balance of the target address to another address. These rights of these operations MUST be allowed only to Issuers.

## Rationale

There are currently no token standards that expressly facilitate conformity to securities law and related regulations. EIP-1404 (Simple Restricted Token Standard) it’s not enough to address FINMA requirements around re-issuing securities to Investors.
In Swiss law, an issuer must eventually enforce the restrictions of their token transfer with a “freeze” function. The token must be “revocable”, and we need to apply a white-list method for AML/KYC checks.

## Backwards Compatibility

> This EIP does not introduce backward incompatibilities and is backward compatible with the older ERC-20 token standard.

Reviewer: It is probably worth mentioning in this section that any application that supports ERC20 tokens may expect transfers to always work.  If the application is not on the appropriate whitelist or is on the frozenlist the application could malfunction.  While technically this is compatible with ERC-20, I think pragmatically a *lot* of stuff will break if a token that implements this is introduced into the ecosystem and that should be mentioned here.

This standard allows the implementation of ERC-20 functions transfer, transferFrom, approve and allowance alongside to make a token fully compatible with ERC-20.

> The token MAY implement decimals() for backward compatibility with ERC-20. If implemented, it MUST always return 0.

Reviewer: Why must decimals be `0`?  If this is a hard requirement, it should probably be mentioned in the specification.

## Implementations

A public Github repository will be added soon.

## References

> ### Standards
> 
> 1. ERC-20 Token Standard. <https://eips.ethereum.org/EIPS/eip-20>
> 

Reviewer: ```suggestion
```
No need for this, inline links to EIPs is the preferred mechanism for cross linking (see suggestions above).

### Swiss Law

1. Federal Act on Combating Money Laundering and Terrorist Financing (Anti-Money Laundering Act, AMLA). <https://www.admin.ch/opc/en/classified-compilation/19970427/202002180000/955.0.pdf>
2. Federal Act on Collective Investment Schemes (Collective Investment Schemes Act, CISA). <https://www.admin.ch/opc/en/classified-compilation/20052154/201607010000/951.31.pdf>

> 3. FINMA publishes ICO guidelines. <https://www.finma.ch/en/news/2018/02/20180216-mm-ico-wegleitung>

Reviewer: Can these be added to `../assets/eip-2980/whatever.pdf` rather than as external links?  In general, we try to avoid external links whenever possible in favor of either inlined content or content that is part of this repository.

> 

Reviewer: This is a non-standard section, and while it isn't a strict requirement to have all sections match the template sections, we generally try to avoid introducing non-standard sections whenever possible.  Can we perhaps inline these links in the Motivation section?  Alternatively, can we create a `### Swiss Law` subsection in the `## Motivation` section instead of adding a new top level section?
